### PR TITLE
Improve multi-image support for STM32 UART and timer drivers

### PR DESCRIPTION
**Pull request by duda-patryk**
Hi!

These changes add UART and timer reset using RCC. This is needed in multi-image environment to make sure that the devices are initialized correctly and won't cause system crash or hang.

Best regards,
Patryk

### CODE REVIEW

**Review by dcpleung — DISMISSED** [review]

**Comment by semihalf-duda-patryk** [comment]
Friendly ping.

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -1615,7 +1615,8 @@ static int uart_stm32_init(const struct device *dev)
 		return err;
 	}
 
-	LL_USART_Disable(config->usart);
+	/* Reset UART to default state using RCC */
+	LL_USART_DeInit(config->usart);
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:1619]: These are 2 different things. Removing `LL_USART_Disable` may be valid. But this should be done in a follow up commit.

File: drivers/counter/counter_ll_stm32_timer.c [hunk 2]
```diff
@@ -470,6 +470,9 @@ static int counter_stm32_init_timer(const struct device *dev)
 	}
 	data->freq = tim_clk / (cfg->prescaler + 1U);
 
+	/* Reset timer to default state using RCC */
+	LL_TIM_DeInit(timer);
```
Comment (erwango) [inline drivers/counter/counter_ll_stm32_timer.c:474]: If we're to add device reset, which will eventually end up in most drivers, I think the clean way is to add include/zephyr/drivers/reset.h support.
I acknowledge this is a significantly larger change though...

**Review by erwango — CHANGES_REQUESTED** [review]
I'd prefer we directly aim for the reset API implementation

File: drivers/serial/uart_stm32.c (hunk 1 above)
Comment (semihalf-duda-patryk) [reply drivers/serial/uart_stm32.c:1619]: Removing `LL_USART_Disable` is valid because after `LL_USART_DeInit` all UART registers have default values. Moved removing `LL_USART_Disable` to separate commit.

File: drivers/counter/counter_ll_stm32_timer.c (hunk 2 above)
Comment (semihalf-duda-patryk) [reply drivers/counter/counter_ll_stm32_timer.c:474]: Could you confirm that I understand it correctly:
- I need to add reset driver for STM32 and modify DTS for all STM32 devices appropriately.
- In the driver I will use RCC to reset requested module.
- Use reset api in UART and timer (and modify their DTS, I guess).

I agree that this approach is much better, cleaner in a long term. If I have any questions I'll let you know.

Comment (erwango) [reply drivers/counter/counter_ll_stm32_timer.c:474]: Your understanding is correct. This is a significantly larger change (but relatively simple).

Comment (semihalf-duda-patryk) [reply drivers/counter/counter_ll_stm32_timer.c]: I added reset drivers and used reset API in UART and timer drivers. It was **significant** amount of work. I wouldn't say it was simple, but definitely worth it.

**Comment by semihalf-duda-patryk** [comment]
Added reset driver for STM32 and STM32MP1. Used reset API in UART and timer drivers.

File: include/zephyr/dt-bindings/reset/stm32f4_reset.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#define STM32_RESET_GPIOI    STM32_RESET(AHB1, 8U)
+#define STM32_RESET_GPIOJ    STM32_RESET(AHB1, 9U)
+#define STM32_RESET_GPIOK    STM32_RESET(AHB1, 10U)
+#define STM32_RESET_CRC	     STM32_RESET(AHB1, 12U)
+#define STM32_RESET_DMA1     STM32_RESET(AHB1, 21U)
+#define STM32_RESET_DMA2     STM32_RESET(AHB1, 22U)
+#define STM32_RESET_DMA2D    STM32_RESET(AHB1, 23U)
+#define STM32_RESET_ETHMAC   STM32_RESET(AHB1, 25U)
+#define STM32_RESET_OTGHS    STM32_RESET(AHB1, 29U)
+#define STM32_RESET_RNG_AHB1 STM32_RESET(AHB1, 31U)
```
Comment (erwango) [inline include/zephyr/dt-bindings/reset/stm32f4_reset.h:37]: Rather than providing this header, it would be easier to provide direct implementation in .dsti files:
```
			gpioa: gpio@40020000 {
				compatible = "st,stm32-gpio";
				gpio-controller;
				#gpio-cells = <2>;
				reg = <0x40020000 0x400>;
				clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00000001>;
				reset = <&rcc STM32_RESET(AHB1, 0U)>;
			};
```
which is more in line with what is done for clocks property and prevents the need to have SoC specific files (stm32f413_reset.h, stm32f446_reset.h).

Comment (semihalf-duda-patryk) [reply include/zephyr/dt-bindings/reset/stm32f4_reset.h:37]: I took GD32 approach as example (eg. include/zephyr/dt-bindings/reset/gd32e10x.h). I will follow your suggestion and change it.

Comment (semihalf-duda-patryk) [reply include/zephyr/dt-bindings/reset/stm32f4_reset.h:37]: Besides of this, does the code look good for you?

**Review by erwango — CHANGES_REQUESTED** [review]
Thanks for this great addition. One request regarding the device tree definition.

File: include/zephyr/dt-bindings/reset/stm32f4_reset.h (hunk 3 above)
Comment (erwango) [reply include/zephyr/dt-bindings/reset/stm32f4_reset.h:37]: Still looking into it, but this was the most visible issue I saw.
Otherwise the remaining part looks good, but I wanted to warn earlier on that part (I wish I could have warn before you perform the complete device tree completion).
Then, it would be nice to propose a specific PR for this whole development which is much boarder than "Improve multi-image support for STM32 UART and timer drivers"

**Comment by semihalf-duda-patryk** [comment]
I replaced defines in DTS with `STM32_RESET()`. To do that, I used magic power of C preprocessor and sed.

**Comment by semihalf-duda-patryk** [comment]
Added missing newline at the end of include/zephyr/dt-bindings/reset/stm32u5_reset.h file.

File: drivers/reset/reset_stm32mp1.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2022 Google Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32mp1_rcc_rctl
```
Comment (erwango) [inline drivers/reset/reset_stm32mp1.c:7]: Instead of duplicating the whole driver, can we think about a (n optional) property than could be use to differentiate behaviors?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/reset/reset_stm32mp1.c (hunk 4 above)
Comment (semihalf-duda-patryk) [reply drivers/reset/reset_stm32mp1.c:7]: Yes, it's possible. I would change `reset_stm32_line_deassert()` to support STM32MP1 + some define changes in driver. Do you think it would be cleaner? Do you have a name for the property?

PS I found bug in drivers/reset/reset_stm32mp1.c just now, so this is not final version.

Comment (erwango) [reply drivers/reset/reset_stm32mp1.c:7]: One idea would be to introduce 2 props on all series which would be set-offset/clr-offset with value 5/5 on most series except MP1 where it would be 5/17.

Comment (semihalf-duda-patryk) [reply drivers/reset/reset_stm32mp1.c:7]: To deassert reset on STM32MP1 you need to set but in clear register (clearing bit in set/clear register does nothing), so we actually need 3 properties. What do you think about it? Can I use default value for set-offset/clr-offset, and add these properties in stm32mp1.dtsi?

Comment (erwango) [reply drivers/reset/reset_stm32mp1.c:7]: Alternative is to use the following for MP1:
```
         compatible = "st,stm32mp1-rcc-rctl", "st,stm32-rcc-rctl";
```
This way you get both compatibles defined when compiled for MP1 case.
Then you can use `DT_HAS_COMPAT_ENABLED(st_stm32mp1_rcc_rct)` to tweak stm32 reset driver behavior in MP1 case. Similar trick is used in SPI and I2C STM32 drivers.

Comment (semihalf-duda-patryk) [reply drivers/reset/reset_stm32mp1.c]: I added `set-bit-to-deassert` property. It's used in MP1 `rctl` node to indicate that the driver should set bit in clear register to deassert reset.

File: dts/bindings/reset/st,stm32-rcc-rctl.yaml [hunk 5]
```diff
@@ -27,5 +27,11 @@ properties:
     "#reset-cells":
       const: 1
 
+    set-bit-to-deassert:
+      type: boolean
+      description: |
+        Indicates if the drives should set bit in clear register to
```
Comment (erwango) [inline dts/bindings/reset/st,stm32-rcc-rctl.yaml:33]:
```suggestion
        Indicates if the driver should set bit in clear register to
```

File: dts/arm/st/f0/stm32f0.dtsi [hunk 6]
```diff
@@ -231,6 +231,7 @@
 			compatible = "st,stm32-timers";
 			reg = <0x40012c00 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000800>;
+			resets = <&rctl STM32_RESET(APB2, 11U)>;
```
Comment (erwango) [inline dts/arm/st/f0/stm32f0.dtsi:236]: dts changes should not be in this commit (drivers: counter: Reset timer using RCC before initialization)

File: dts/arm/st/f0/stm32f0.dtsi [hunk 7]
```diff
@@ -159,6 +159,7 @@
 			compatible = "st,stm32-usart", "st,stm32-uart";
 			reg = <0x40013800 0x400>;
 			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00004000>;
+			resets = <&rctl STM32_RESET(APB2, 14U)>;
```
Comment (erwango) [inline dts/arm/st/f0/stm32f0.dtsi:162]: Ditto. dts changes don't belong to this commit

**Review by erwango — APPROVED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
wrong button: some changes are actually required

**Comment by semihalf-duda-patryk** [comment]
Split changes in DTS from changes in driver. Fixed typo.

File: dts/arm/st/f0/stm32f0.dtsi (hunk 6 above)
Comment (semihalf-duda-patryk) [reply dts/arm/st/f0/stm32f0.dtsi:236]: Fixed.

**Review by erwango — APPROVED** [review]

File: drivers/serial/uart_stm32.c [hunk 8]
```diff
@@ -1618,7 +1619,13 @@ static int uart_stm32_init(const struct device *dev)
 		return err;
 	}
 
-	LL_USART_Disable(config->usart);
```
Comment (erwango) [inline drivers/serial/uart_stm32.c:1621]: This should be kept.
On some series (H7 at least),  device shjould be disabled to set some registers such as baudrate

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c (hunk 8 above)
Comment (semihalf-duda-patryk) [reply drivers/serial/uart_stm32.c:1621]: I think RCC reset changes UE bit (Uart Enable) to 0, but I can keep it to be extra safe.

**Comment by semihalf-duda-patryk** [comment]
Restored call to `LL_UART_Disable`

File: drivers/serial/uart_stm32.c (hunk 8 above)
Comment (erwango) [reply drivers/serial/uart_stm32.c:1621]: ok right. That should be easy to check anyway. But let's keep that for an extra PR if needed.

**Review by erwango — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]
